### ISSUE TITLE
No ReadOnlySpan<byte> static data optimization for new byte[N]

### ISSUE THREAD

**Issue by bgrainger** [dotnet/roslyn#33088]
**Version Used**: dotnet 2.2.200-preview-009748

**Steps to Reproduce**:

Declare a static `ReadOnlySpan<byte>` property of the form:

```csharp
public class C
{
    public static ReadOnlySpan<byte> StaticData => new byte[] { 0 };
    public static ReadOnlySpan<byte> DoesntWork => new byte[1];
} 
```

**Expected Behavior**:

Per the optimisation in #24621, I expected that both `StaticData` and `DoesntWork` would be compiled to a reference to static data, e.g., `new ReadOnlySpan<byte>(&global::<PrivateImplementationDetails>.5BA93C9DB0CFF93F52B521D7420E43F6EDA2784F, 1);`

**Actual Behavior**:

Only `StaticData` has the implementation optimisation; `DoesntWork` doesn't, even though it has the exact same data. The problem is more noticeable with an expression like `new byte[1000]`, where you might not want to type out 1000 literal zero bytes in source code.

SharpLab: https://sharplab.io/#v2:D4AQDABCCMDcCwAoJIDMUBMEDCSDeSERU6MAbBAEoCmAhgCYDyAdgDYCeAygA63MA8AI3YAXagD4InEbREBLAMYARWbQgBeSc2oB3CMLEBtALoQ8ESAF8EiYiSjQKNBiw48+Q0RIhKA9tQBnZhEAdV8AJwBrDS1dfS9DaGMbSwggA===

**Comment by gafter** [comment]
If I'm not mistaken, @alrz did the original optimization.

**Comment by alrz** [comment]
@gafter I don't think this is related though. I did this for stackalloc but we had to revert it eventually because some endianness issues with initblk.

I this the problem here is that #24621 only works if there's an array initializer, otherwise it allocate and wraps the array. 

tagging @VSadov

**Comment by canton7** [comment]
Would we want to use static data on the case of a large all-zero array? Unless I'm mistaken, that's trading a possible runtime allocation for a definite increase is assembly size.

If we care about optimising this specific case, wouldn't a singleton byte array be better? It's only allocated if it's actually needed, and doesn't increase the assembly size.

**Comment by bgrainger** [comment]
Note: `new byte[1000]` in my OP was an exaggeration; my actual use case was `new byte[19]`: https://github.com/mysql-net/MySqlConnector/blob/0dbdca44eab2b7c967fbbb72c734f4dd9a59952c/src/MySqlConnector/Protocol/Payloads/HandshakeResponse41Payload.cs#L71-L72

In this case, I thought the tradeoff (assembly size vs allocation) was worth it, but you're right that it's probably not a good idea (in general) to store large arrays of zeroes in the assembly itself.

**Comment by nbalakin** [comment]
I implemented a bunch of unit tests to cover the both #24621 optimization and the new one and we have a problem here. The optimization implemented in #24621 generates a code that violates ECMA-355 verifiability rules.

Here is the code generated:

```
ldsflda    int16 '<PrivateImplementationDetails>'::'__field_name__'
ldc.i4.2
newobj     instance void valuetype System.ReadOnlySpan`1<uint8>::.ctor(void*, int32)
ret
```

Here is what ECMA-355 says in III.4.15:

> ldsflda field
> ...
> Verifiability:
> For verifiable code, field cannot be init-only.

Is intending violation of the rule? If so I'll just document the violation with a comment.
Or should I fix the violation?

**Comment by nbalakin** [comment]
I suppose another ctor can be used (`ctor(T?[])`). That allows to verify at least user assemblies.

**Comment by agocke** [comment]
I'm not really sure how this is an optimization. It's just trading binary size for allocations, isn't it?

**Comment by nbalakin** [comment]
I'd agree with considering the optimization as a controversial one... But I've found an issue with "help-wanted" label and why not to implement it. 🙂


> The optimization implemented in #24621 generates a code that violates ECMA-355 verifiability rules.

I've dug up into the code and found it's intended and the code generated is verifiable if it's asked by turning on appropriate feature toggle.

I've created a new PR with the old optimization unit tests only just to cover the existing code.

**Comment by agocke** [comment]
Yeah, I'm not sure the full consequences were appreciated when the issue was tagged. The original is definitely an optimization because the actual values in the initializer *must* be present in the binary, so that code just avoids an extra allocation.

This is more of a tradeoff and I think most people would much rather just allocate an empty block of memory, as opposed to making the binary bigger.

@gafter did you have a different opinion?

**Comment by VSadov** [comment]
I think `new T[]` case was just overlooked. 
The following means basically the same.

```cs
    public static ReadOnlySpan<byte> StaticData => new byte[] { 0, 0, 0 };
    public static ReadOnlySpan<byte> DoesntWork => new byte[3];
```

Readonly spans of zeroes are probably not very common and user can do `new T[] {0,0,0}`, if needed. 
I think optimization could be extended to cover `new T[]` case just for consistency. 

Assuming it is not common, it is probably not a big deal one way or another.

**Comment by agocke** [comment]
Thanks @VSadov. I suppose the problem with making those the same thing is that you can't choose *not* to do the optimization. What would the suggestion be to defeat it? Store the array in a local first?

**Comment by agocke** [comment]
I'll leave this one up to @gafter since I can see both an argument for consistency, but it also seems to be a trade-off that isn't directly in the "optimization" category.
